lib/Pipes/TaggedFunctionKind.cpp

```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+                                  const GlobalTupleTreeDiff &Diff,
+                                  const pipeline::Global &Before,
+                                  const pipeline::Global &After) const {
+  using PM = PathMatcher;
+  using Binary = model::Binary;
+  if (not llvm::isa<TupleTreeGlobal<Binary>>(Before))
+    return;
+
+  const auto &OldModel = cast<TupleTreeGlobal<Binary>>(Before).get();
+  const auto &Model = cast<TupleTreeGlobal<Binary>>(After).get();
+
+  // If CrossRelations is not found, the most conservative approach is to
+  // invalidate all targets.
+  if (not Ctx.containsReadOnlyContainer(BinaryCrossRelationsRole)) {
+    ToRemove.emplace_back(Target(*this));
+    return;
+  }
+
+  static constexpr auto &BCRR = BinaryCrossRelationsRole;
+  using CRContainer = TupleTreeContainer<CrossRelations>;
+  const auto &Container = Ctx.getReadOnlyContainer<CRContainer>(BCRR);
+  const TupleTree<CrossRelations> &CrossRelations = Container.get();
+
+  const TupleTreeDiff<Binary> *ModelDiff = Diff.getAs<Binary>();
+  if (not ModelDiff)
+    return;
+
+  auto AttrMatcher = PM::create<Binary>("/Functions/*/Attributes").value();
+
+  std::set<std::string> Targets;
+  const auto &IDF = Model->ImportedDynamicFunctions;
 
-  if (not Ctx.containsReadOnlyContainer(BinaryCrossRelationsRole))
+  for (const Change<Binary> &Change : ModelDiff->Changes) {
+    revng_assert(not Change.Path.empty());
+
+    // A new function has been added or removed?
+    if (pathAsString<Binary>(Change.Path) == "/Functions") {
+      if (Change.New) {
+        // Invalidate all targets
+        for (const model::Function &Function : Model->Functions)
+          Targets.insert(Function.Entry.toString());
+      } else if (Change.Old) {
+        // Invalidate the function and its callers
+        auto Function = std::get<model::Function>(*Change.Old);
+        Targets.insert(Function.Entry.toString());
+        insertCallers(CrossRelations, Function.Entry, Targets);
+      }
+    }
+
+    // A new prototype has been added or removed?
+    if (pathAsString<Binary>(Change.Path) == "/Types") {
+      if (Change.New) {
+        // Invalidate all targets
+        for (const model::Function &Function : Model->Functions)
+          Targets.insert(Function.Entry.toString());
+      } else if (Change.Old) {
+        auto Prototype = std::get<model::TypePath>(*Change.Old);
+
+        for (const model::Function &Function : Model->Functions) {
+          if (Prototype == Function.Prototype) {
+            // Invalidate the function and its callers
+            Targets.insert(Function.Entry.toString());
+            insertCallers(CrossRelations, Function.Entry, Targets);
+          }
+        }
+
+        for (const model::DynamicFunction &Function : IDF) {
+          if (Prototype == Function.Prototype) {
+            // Invalidate callers of the model::DynamicFunction
+            insertCallers(CrossRelations, Function.OriginalName, Targets);
+          }
+        }
```

Comment:
Given what we do below, I don't think this is necessary.
Even if the type is completely changed, as long as the FSO and set of preserved registers is the same, we don't need to invalidate.

Reply:
Uhmm, right, good catch, thanks. I drop it here, and likely reuse it for `ABIEnforced` Kind.